### DISCUSSION TITLE
Bug in contour-plot with array channels and inset

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [observablehq/plot#1993] [Q&A]
There seems to be a bug in the contour-plot when used with array-channels and an inset, where instead of having an empty space for the inset, the edges of the contour-plot get smeared, plus other artifacts.

Please have a look at this [Notebook](https://observablehq.com/d/7655505d29e8c517).

Thanks!

![Bug contour-plot and inset](https://github.com/observablehq/plot/assets/13588114/042a072b-cab3-43d2-8dfc-1230b72b0a55)

**Comment by Fil** [comment]
It's not a bug, it's just a different chart than what you expect. Try this option for example:
>        interpolate: "barycentric"

The two charts are different because they map different things: at the top, it's samples at certain locations, with associated values, and we're trying to interpolate the value across the whole chart. At the bottom, it's a regular array of pixels, all have a given value and there's no interpolation (or extrapolation for that matter).

To clarify the role of extrapolation: when we _extrapolate_ values defined from a straight-line edge, we expect to get this "smearing" pattern.

**Reply by Hvass-Labs** [reply]
Thanks for the explanation.

However, according to your [docs](https://observablehq.com/plot/features/scales#position-scale-options): *"The inset scale options can provide “breathing room” to separate marks from axes or the plot’s edge."* So my expectation is that when I make an inset there should be white space between the marks and the frame, and not this smearing effect.

The reason I want a small inset in these plots, is because I color the background green/red for pos/neg values as a helpful indicator to the viewer.

I tried using `interpolate: "barycentric"` as you suggested and I got the following results. It is the exact same plotting code in both, it is just the data that is different. The first plot looks as I would expect with an exaggerated inset of 100, but the second plot is weirdly smeared. It looks like the contour-plot doesn't always respect the inset properly?

These also show the problem with the tip-mark discussed in another thread - just to rub it in :-)

Other than these issues, the plots look really good, and they are super-fast to generate!

![Bug contour-plot and inset (2)](https://github.com/observablehq/plot/assets/13588114/5e5d4565-5291-4b1e-8fe7-2eca55d98df6)

![Bug contour-plot and inset (3)](https://github.com/observablehq/plot/assets/13588114/fb722a02-aa87-406d-a026-5617d5adfea6)

**Reply by Fil** [reply]
Please share actual code and data that reproduces the issue; we can't continue with these guessing games.

**Reply by Hvass-Labs** [reply]
The code is already provided in the [Notebook](https://observablehq.com/d/7655505d29e8c517) linked at the top. I have added the data for the second plot to it now, if that will help you debug this problem. It is obviously a bug in your code, I was just trying to be polite and constructive. Your testy remark is not appropriate as I am spending a lot of time and effort helping you locate problems in *your* code. A modicum of gratitude would be in order.

**Reply by Fil** [reply]
As I outlined in my answer above, the code in the two charts is different; in fact they see the same data as very different objects, which is why they use different algorithms to plot it—and result in very different images. "Exact same plotting code in both" doesn't match the code I'm seeing in the notebook. In addition, the notebook does not match the image you're sharing. All I'm saying is that it's really difficult to understand an issue without a clear way to reproduce it.

**Reply by Hvass-Labs** [reply]
Perhaps we're talking past each other. Let me briefly explain my setup: I have originally made this on a web-site, and it is really hard to give you the original code, because I have actually made a whole "meta" plotting library on top of your library. So I distilled the essential plot-code into the Notebook above, which shows two plots: The first plot uses array-channels and has the problem I am describing. The second plot does not use array-channels and looks fine. But you can forget about the second plot if that is confusing.

The Notebook now contains two data-sets `data.json` and `data2.json` that can be switched in this line:

    data = FileAttachment("data.json").json()
    data = FileAttachment("data2.json").json()

The images below show the results. One is shown correctly with a very wide inset, while the other is smearing the contour-plot across the inset. This looks like a bug to me. Because according to your docs (quoted above), the purpose of the inset is to provide empty space between the plot-frame and the mark.

If this code and data is not enough for you to debug this problem, then let me know what you need.

![Bug contour-plot and inset (4)](https://github.com/observablehq/plot/assets/13588114/80d258f6-da7f-447f-a518-30d8fd8373c1)

![Bug contour-plot and inset (5)](https://github.com/observablehq/plot/assets/13588114/6844fb69-9930-4c64-b0fb-0794c1f16910)
